go/vt/vtgate/evalengine/fn_string.go

```diff
@@ -127,6 +139,266 @@ var _ IR = (*builtinLeftRight)(nil)
+	// MySQL converts all the arguments to DOUBLE
+	containsOnlyInt64, containsOnlyString := true, true
+	for _, arg := range args {
+		if arg == nil {
+			continue
+		}
+
+		containsOnlyInt64 = sqltypes.IsIntegral(arg.SQLType()) && containsOnlyInt64
+		containsOnlyString = !sqltypes.IsNumber(arg.SQLType()) && containsOnlyString
+	}
+
+	if containsOnlyInt64 {
+		tar := evalToInt64(args[0])
+
+		for i, arg := range args[1:] {
+			if arg == nil {
+				continue
+			}
+
+			e := evalToInt64(arg)
+			if tar.i == e.i {
+				return newEvalInt64(int64(i + 1)), nil
+			}
+		}
+	} else if containsOnlyString {
+		tar, ok := args[0].(*evalBytes)
+		if !ok {
+			tar, err = evalToVarchar(args[0], call.collate, true)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		for i, arg := range args[1:] {
+			if arg == nil {
+				continue
+			}
+
+			var ok bool
+			e, ok := arg.(*evalBytes)
+			if !ok {
+				e, err = evalToVarchar(arg, call.collate, true)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			// Compare target and current string
+			if len(tar.bytes) == len(e.bytes) {
+				eq := true
+				for i, b := range tar.bytes {
+					if e.bytes[i] != b {
+						eq = false
+						break
+					}
+				}
+
+				if eq {
+					return newEvalInt64(int64(i + 1)), nil
+				}
+			}
+		}
+	} else {
+		tar, _ := evalToFloat(args[0])
+
+		for i, arg := range args[1:] {
+			if arg == nil {
+				continue
+			}
+
+			e, _ := evalToFloat(arg)
+			if tar.f == e.f {
+				return newEvalInt64(int64(i + 1)), nil
+			}
+		}
+	}
+
+	return newEvalInt64(0), nil
+}
+
+func (call *builtinField) compile(c *compiler) (ctype, error) {
+	strs := make([]ctype, len(call.Arguments))
+
+	for i, arg := range call.Arguments {
+		var err error
+		strs[i], err = arg.compile(c)
+		if err != nil {
+			return ctype{}, err
+		}
+	}
+
+	// If the arguments contain both integral and string values
+	// MySQL converts all the arguments to DOUBLE
+	containsOnlyString, containsOnlyInt64 := true, true
+	for _, str := range strs {
+		if sqltypes.IsNull(str.Type) {
+			continue
+		}
+
+		containsOnlyString = !sqltypes.IsNumber(str.Type) && containsOnlyString
+		containsOnlyInt64 = sqltypes.IsIntegral(str.Type) && containsOnlyInt64
+	}
+
+	if containsOnlyInt64 {
+		for i, str := range strs {
+			offset := len(strs) - i
+			skip := c.compileNullCheckOffset(str, offset)
+
+			_ = c.compileToInt64(str, offset)
+			c.asm.jumpDestination(skip)
+		}
+
+		c.asm.Fn_FIELD_i(len(call.Arguments))
+	} else if containsOnlyString {
+		for i, str := range strs {
+			offset := len(strs) - i
+			skip := c.compileNullCheckOffset(str, offset)
+
+			switch {
+			case str.isTextual():
+			default:
+				c.asm.Convert_xce(offset, sqltypes.VarChar, call.collate)
+			}
+			c.asm.jumpDestination(skip)
+		}
+
+		c.asm.Fn_FIELD_b(len(call.Arguments))
+	} else {
+		for i, str := range strs {
+			offset := len(strs) - i
+			skip := c.compileNullCheckOffset(str, offset)
+
+			c.asm.Convert_xf(offset)
```

Comment:
I think we can guard this with a check if it's not already a float? So we don't have to emit the conversion in that case?